syntactic/src/main/scala/syntactic_checker/whitelist/Feature.scala

```diff
@@ -0,0 +1,60 @@
+package syntactic_checker.whitelist
+
+import scala.meta.Tree
+
+/**
+ * Subdivision of the Scala language, allowing a specific set of constructs
+ */
+sealed trait Feature {
+
+  /**
+   * Checks whether a tree is allowed by this feature
+   *
+   * @param tree the tree to check
+   * @return true iff it is allowed
+   */
+  def allows(tree: Tree): Boolean
+
+}
+
+object Feature {
+
+  /**
+   * Feature defined as a combination of other features
+   * A composite feature allows a tree iff at least one of the underlying features accepts it
+   *
+   * @param features the features to allow
+   */
+  abstract class CompositeFeature(private val features: Set[Feature]) extends Feature {
+    require(features != null)
+
+    def this(features: List[Feature]) = this(features.toSet)
+
+    def this(features: Feature*) = this(features.toSet)
+
+    override def allows(tree: Tree): Boolean = features.exists(_.allows(tree))
+
+    override def equals(other: Any): Boolean = other match {
+      case that: CompositeFeature =>
+        features == that.features
+      case _ => false
+    }
+
+    override def hashCode(): Int = features.hashCode()
```

Comment:
This is relying on equality of `Feature`, which can be arbitrarily defined since end-users can extend `AtomicFeature`.

Reply:
Fixed in b6550557a6c78e709b1fdc5323385cd5df84f7bb

Reply:
But we still rely on equality to be well defined in some other places (e.g. when we use a `Set[Feature]`, and more generally when we add/remove features from a collection of features).

The only way I can think of to have reliable equality would be to make the class `AtomicFeature` sealed. But that means end-users wouldn’t be able to define their own features. Another possibility is to document that `AtomicFeature` should be only extended by singleton `object`s.

Reply:
I found online that using `self: Singleton =>` ensures that the class cannot be extended by something else than an object.
Would this work?
```Scala
abstract class AtomicFeature(checkPF: PartialFunction[Tree, Boolean]) extends Feature { self: Singleton =>
    require(checkPF != null)

    override def allows(tree: Tree): Boolean = {
      checkPF.applyOrElse(tree, (_: Tree) => false)
    }
  }
```
(I would need to document it too to avoid confusion, of course)

Reply:
Not sure this really works in practice, I guess anyone could write a `class Foo extends Singleton`. I think just documenting is enough.

Reply:
OK, I simply documented it (and I also documented that `BlacklistRule` should only be extended by objects)

(7d419b3)